dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,243 @@
+from __future__ import unicode_literals
+
+import os
+import posixpath
+import logging
+
+from funcy import cached_property
+from backoff import on_exception, expo
+
+from dvc.scheme import Schemes
+from dvc.path_info import CloudURLInfo
+from dvc.remote.base import RemoteBASE
+from dvc.config import Config
+from dvc.exceptions import DvcException
+from dvc.remote.gdrive.pydrive import (
+    RequestListFile,
+    RequestListFilePaginated,
+    RequestCreateFolder,
+    RequestUploadFile,
+    RequestDownloadFile,
+)
+from dvc.remote.gdrive.utils import FOLDER_MIME_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class GDriveURLInfo(CloudURLInfo):
+    @property
+    def netloc(self):
+        return self.parsed.netloc
+
+
+class RemoteGDrive(RemoteBASE):
+    scheme = Schemes.GDRIVE
+    path_cls = GDriveURLInfo
+    REGEX = r"^gdrive://.*$"
+    REQUIRES = {"pydrive": "pydrive"}
+    GDRIVE_USER_CREDENTIALS_DATA = "GDRIVE_USER_CREDENTIALS_DATA"
+    CREDENTIALS_FILE_PATH = "credentials.json"
+
+    def __init__(self, repo, config):
+        super(RemoteGDrive, self).__init__(repo, config)
+        self.no_traverse = False
+        self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        self.config = config
+        self.init_drive()
+
+    def init_drive(self):
+        self.gdrive_credentials_path = self.config.get(
+            Config.SECTION_REMOTE_KEY_FILE, None
+        )
+        if not self.gdrive_credentials_path:
+            raise DvcException(
+                "Google Drive settings file path is missed from config. "
+                "Learn more at "
+                "https://man.dvc.org/remote/add."
+            )
+        self.root_id = self.get_path_id(self.path_info, create=True)
+        self.cached_dirs, self.cached_ids = self.cache_root_dirs()
+
+    @on_exception(expo, DvcException, max_tries=8)
```

Comment:
> Doing it 8 times with unknown exponential backoff.

Number `8` is just first what I've tried (seems, suggested in decorator's usage sample). So, yes, need to try to decrease it and check when it will start fail. The results might vary because of hardware and connection speed.

> We are retrying on all dvc exceptions

Right, I will introduce new exception type specifically to be expected by decorator.

> Is there any reason it should be done for GDrive separately from all the other remotes? Is it somehow exceptionally unreliable?

Yes, requests with 403 and 50X errors should be refreshed ( https://developers.google.com/drive/api/v2/manage-uploads#errors )
btw, PyDrive uses internally google-api-python-client which already has default retry logic for above error codes, but sometimes it's not enough (in case of 403 User Limits Exceeded its always not enough :) ).

Reply:
https://github.com/googleapis/google-api-python-client/commit/c6425a00e3c16268b101703c45f2a3e53154a372#diff-84a6c030bfdfe3ce2eac14a3369e6114R107

Reply:
By arbitrary I also mean that the base, power and cap are unknown. Does it wait for milliseconds or days you can't know from this code, unless without additional digging.

We also already use [funcy.retry](https://funcy.readthedocs.io/en/stable/flow.html#retry), which makes this very obvious:
```python
# 8 tries, start at 0.5s, multiply by golden ratio, cap at 10s
@retry(8, 
       errors=(ErrorClass, OtherErrorClass), 
       timeout=lambda a: min(0.5 * 1.618 ** a, 10))
def some_func(...):
    # ... 
```

Reply:
👍 @MaxRis makes sense to use fancy if it's already a dependency and it's more explicit

Reply:
thanks, `funcy.retry` looks much better!

Reply:
@MaxRis @Suor can be resolved?

Reply:
Resolving this conversation, thanks